dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp

```diff
@@ -127,15 +148,92 @@ bool PredicateExpressionsOptimizer::allowPushDown(const ASTSelectQuery * subquer
         {
             const auto & function = FunctionFactory::instance().tryGet(subquery_function->name, context);
 
-            /// Skip lambda、tuple and other special functions
+            /// Skip lambda, tuple and other special functions
             if (function && function->isStateful())
                 return false;
         }
+    }
+
+    const auto * ast_join = ast_select->join();
+    const ASTTableExpression * left_table_expr = nullptr;
+    const ASTTableExpression * right_table_expr = nullptr;
+    const ASTSelectQuery * left_subquery = nullptr;
+    const ASTSelectQuery * right_subquery = nullptr;
 
-        return true;
+    if (ast_join)
+    {
+        left_table_expr = ast_select
+                                ->tables()->as<ASTTablesInSelectQuery>()
+                                ->children[0]->as<ASTTablesInSelectQueryElement>()
+                                ->table_expression->as<ASTTableExpression>();
+        right_table_expr = ast_select
+                                ->tables()->as<ASTTablesInSelectQuery>()
+                                ->children[1]->as<ASTTablesInSelectQueryElement>()
+                                ->table_expression->as<ASTTableExpression>();
+
+        if (left_table_expr && left_table_expr->subquery)
+            left_subquery = left_table_expr->subquery
+                                ->children[0]->as<ASTSelectWithUnionQuery>()
+                                ->list_of_selects->children[0]->as<ASTSelectQuery>();
+        if (right_table_expr && right_table_expr->subquery)
+            right_subquery = right_table_expr->subquery
+                                ->children[0]->as<ASTSelectWithUnionQuery>()
+                                ->list_of_selects->children[0]->as<ASTSelectQuery>();
+
+        /// NOTE: the syntactic way of pushdown has limitations and should be partially disabled in case of JOINs.
+        ///       Let's take a look at the query:
+        ///
+        ///           SELECT a, b FROM (SELECT 1 AS a) ANY LEFT JOIN (SELECT 1 AS a, 1 AS b) USING (a) WHERE b = 0
+        ///
+        ///       The result is empty - without pushdown. But the pushdown tends to modify it in this way:
+        ///
+        ///           SELECT a, b FROM (SELECT 1 AS a) ANY LEFT JOIN (SELECT 1 AS a, 1 AS b WHERE b = 0) USING (a) WHERE b = 0
+        ///
+        ///       That leads to the empty result in the right subquery and changes the whole outcome to (1, 0) or (1, NULL).
```

Comment:
👍 Are we going to move the entire predicate optimization onto the execution stack in the future ? and how do we do that ? Maybe I can get on with this work?

Reply:
Yes, we are going to do it after "Processors" will be ready https://github.com/yandex/ClickHouse/pull/4914
And we will be happy to have your help!